daemon/logger/ring.go

```diff
@@ -0,0 +1,122 @@
+package logger
+
+import (
+	"context"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+)
+
+const (
+	ringCloseTimeout = 10 * time.Second
+	ringSize         = 10000
+)
+
+// RingLogger is a ring buffer that implements the Logger interface.
+// This is used when lossy logging is OK.
+type RingLogger struct {
+	queue  chan Message
+	ctx    context.Context
+	cancel func()
+	l      Logger
+	logCtx Context
+}
+
+type ringWithReader struct {
+	*RingLogger
+}
+
+func (r *ringWithReader) ReadLogs(cfg ReadConfig) *LogWatcher {
+	reader, ok := r.l.(LogReader)
+	if !ok {
+		// something is wrong if we get here
+		panic("expected log reader")
+	}
+	return reader.ReadLogs(cfg)
+}
+
+func newRingLoggerWithSize(driver Logger, logCtx Context, size int) *RingLogger {
+	ctx, cancel := context.WithCancel(context.Background())
+	l := &RingLogger{
+		queue:  make(chan Message, size),
+		ctx:    ctx,
+		cancel: cancel,
+		l:      driver,
+		logCtx: logCtx,
+	}
+	l.run()
+
+	return l
+}
+
+// NewRingLogger creates a new Logger that is implemented as a RingBuffer wrapping
+// the passed in logger.
+func NewRingLogger(driver Logger, logCtx Context) Logger {
+	l := newRingLoggerWithSize(driver, logCtx, ringSize)
+	if _, ok := driver.(LogReader); ok {
+		return &ringWithReader{l}
+	}
+	return l
+}
+
+// Log queues messages into the ring buffer
+func (r *RingLogger) Log(msg *Message) error {
+	var m Message
+	m = *msg
+	select {
+	case r.queue <- m:
+	case <-r.ctx.Done():
+		select {
+		case r.queue <- m:
+		case <-time.After(ringCloseTimeout):
+		}
+	default:
+		select {
+		case <-r.ctx.Done():
+		case <-r.queue: // Make room on the queue
```

Comment:
May need to add a lock to protect against another log message coming in right after this but before we get a chance to queue below.